### ISSUE TITLE
🚀 Feature: Techdocs from confluence

### ISSUE THREAD

**Issue by cb-jarunmadhesh** [backstage/backstage#28203]
### 🔖 Feature description

A new feature that enables Backstage TechDocs to source documentation directly from Confluence.

### 🎤 Context

Many organizations, including my company, rely heavily on Confluence for documentation because it is user-friendly and makes it easy to create and maintain content. Migrating or duplicating content to Markdown-based systems can be a cumbersome and time-consuming process.

Supporting Confluence as a TechDocs source would allow Backstage to better align with enterprise workflows and reduce barriers to adoption. My company uses Confluence for its documentation, and having TechDocs integration for Confluence would be immensely beneficial both for us and for other organizations with similar needs.

While there are existing tools, such as a scaffolder to migrate Confluence docs to Markdown on Github, this approach eliminates the need for manual migration and the challenge of maintaining duplicate documentation across Confluence and TechDocs.

### ✌️ Possible Implementation

The Preparer could be extended to support reading directly from Confluence by introducing a new Confluence URL Reader. This would allow TechDocs to accept and process Confluence URLs directly. 

I have a poc implementation and would be happy to raise a PR if this feature request is approved.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

**Comment by awanlin** [comment]
Hi @cb-jarunmadhesh, On the surface this sounds like a good idea for some, but I worry that this will end up as a poor experience. 

Do you have a working implementation of this? If so can you share screen shots of what this looks like? Will this just bring in a single page or a whole set of pages? Would this require Adopters to follow some specific convention to make it work? How would the navigation work? What happens when pages have plugins being used? How does the HTML look and will it work with existing themes properly?

**Comment by cb-jarunmadhesh** [comment]
Hi @awanlin,
Yes, I understand. Let’s ensure this is implemented properly:

1. Yes, I do have a working example. PFA a screen recording.
2. It'll fetch the page along with its child pages and their attachments. In Confluence, a page can have child pages. While converting to Markdown, I’m creating a Markdown file and a folder with the same name, which will contain the child pages, representing the directory structure.
3. Adopters will need to configure the Confluence integration in app-config and then can add their page URL to their component's `backstage.io/techdocs-ref` annotation. That should do.
4. Currently, I’m not handling cases where a plugin is being used within a page. If that happens, the plugin content will not appear in Backstage doc.
5. HTMLs are still being generated the same way as they are for all other providers. I’ve simply edited the preparer to fetch pages from Confluence and feed them to the generators.


https://github.com/user-attachments/assets/6b19e503-154f-49e4-bc23-f43347389724

**Comment by awanlin** [comment]
Thanks for the follow up @cb-jarunmadhesh, this video helps settle all my worries. This looks like it would be a good experience! 👍 

My reality is much different than yours: most, if not all, the Backstage Adopters I speak to say that tools like Confluence (and Sharepoint) end up being a mess and Engineers never know where to put things. The data we see is that writing documentation along with the code reduces friction for Engineers and results in more documentation being added and maintained.

**Comment by rjoschi** [comment]
Hi, Any update/progress on this? We are interested in this as well.

**Comment by mjramer** [comment]
My org also heavily relies on Confluence for documentation, but as part of our Backstage adoption we're driving a culture shift to docs in Markdown for TechDocs. In addition to @awanlin's point above, we worry engineers will forget to update docs as the code changes. As a bonus, we're able to build a GenAI Docs Update CI/CD job to suggest docs updates.

**Comment by konradsemsch** [comment]
I agree with the point that for developers maintaining docs along with code is the easiest, but there are also non-technical stakeholders, who need to maintain certain product\ feature documentation. And for them the goal to tool is Confluence. So merging the best of both worlds would enable both tech and non-tech users to maintain different types of docs in the tool they prefer.

@cb-jarunmadhesh - did you further work on it? Did you publish this as a custom plugin? I would be very interested in extending this as well.

**Comment by camilaibs** [comment]
Hi there 👋🏻 , I noticed that the pull requests linked to the issue became stale and closed. Would anyone like to resume work on it and help us finalize it? Contributions are more than welcome 🙌🏻 .

**Comment by joelml** [comment]
Hello! 👋 I have been watching this for a while and based on the work done in the original PRs I have managed to create a solution that works using a backend module for the techdocs.

It allows for annotations like:
backstage.io/techdocs-ref: url:https://confluence.yourcompany.com/...

I have never submitted anything to the Backstage community repo before but it would perhaps fit best there(?)
I will try to do this in the next few days 😊

**Comment by joelml** [comment]
Finally created the PR, sorry for the delay!
https://github.com/backstage/community-plugins/pull/6387

**Comment by joelml** [comment]
> Hi there 👋🏻 , I noticed that the pull requests linked to the issue became stale and closed. Would anyone like to resume work on it and help us finalize it? Contributions are more than welcome 🙌🏻 .

@camilaibs The PR has been merged so there is a community plugin now 😄

**Comment by awanlin** [comment]
@joelml might make sense to also get it added to the [Plugin Directory](https://backstage.io/plugins/)? Here's how: https://backstage.io/docs/plugins/add-to-directory

**Comment by joelml** [comment]
@awanlin Yes, I would like to add it there but when searching I can't seem to find the NPM package for it so it's yet to be pushed from the Backstage community plugins. Do you know if I'm to do something there to make it happen? If I have missed something I of course want to fix it.

**Comment by awanlin** [comment]
Isn't this it? https://www.npmjs.com/package/@backstage-community/plugin-techdocs-backend-module-confluence

**Comment by joelml** [comment]
There it is! Search function didn't give me any hits! Thanks for pointing it out, I'll add this to the Plugin Directory shortly

**Comment by awanlin** [comment]
Honestly I sort of assumed we forgot to remove the private flag in the `package.json` and that's a common miss when reviewing PRs for new plugins. When it was not there and I saw there way a changelog file I just grabed the name and popped i in my URL for NPM

https://github.com/backstage/community-plugins/blob/41f738e480dd8e52319287cd7bf48e3712c884a1/workspaces/confluence/plugins/techdocs-backend-module-confluence/package.json#L2C12-L2C74

**Comment by vinzscam** [comment]
Closed by https://github.com/backstage/community-plugins/pull/6387. Thanks!
